docs/proposals/odo-config-file.md

```diff
@@ -0,0 +1,253 @@
+
+IDEs often use a configuration file or directory to hold specific configuration for an application, workspace or the whole tool. These files and directories are stored under the same control version system as where the source code lives, so that anytime a developer checks out that repository has control on how to configure that tool to work with that source code.
+
+Git is another example of a tool that provides configuration, that can be related of every instance of any repository the user has or to a single instance. These are defined as global and local config. Git promoves the fact that local configuration is always more important than global configuration.
+
+## Motivation
+
+Provide a way to preserve conscisous decisions the developer has made that are orthogonal to the code but relevant to the tool/platform. Any user that continues working on an existing application don't need to know the decisions the original developer took, as these are preserved along with the source code and provides for better reproducibility of the development environment setup.
+
+## Use Cases
+
+1. As a user, I want to be able to preserve some configuration I've done about my current component along with my source code
+1. As a user, I want to be able to use configuration stored with my source code to help me be more productive
+1. As a user, I want to be able to manage explicit configuration that is saved with my source code
+1. As a user, I want to be able to check out an application source code on different computers and have them producing the same result in relation to the development environment (deployment of the application on the platform)
+
+## Design overview
+
+We need to provide a configuration file `odo-config` in a configuration directory `.odo`, that will hold specific configuration that `odo` will use on execution. This configuration might modify how the CLI work or how the component will work on the platform.
+
+### Scopes of configuration
+
+`odo` CLI configuration needs to be preserved in configuration files. As a developer might work on multiple
+components, there is a need to have different scopes for configuration so that a developer don't need to
+explicitly configure each and every component independently.
+
+#### Local configuration
+
+Every component can be configured independently. We call this local configuration for the component as it only affects the component it references. This configuration will be stored in a file in `.odo` specific directory located at the root directory of the component's source code. In this way, the configuration file can be added to version control system and preserved along with the application. If any other developer checks out the same source code he will get the same configuration as the original developer.
+
+We propose to have a `.odo` folder in the root of the application's source code directory to store every odo local configurations.
```

Comment:
sorry i might be missing some context here
but why do we need `.odo` folder instead of `.odo` file?
is it because of multiple components under same application? but aren't we assuming a repo per component?_

Reply:
I did send an email around on tuesday and we discuss at cabal. We switched from file to folder so we can have more configuration than what a single file can store. In the future this might be fragments, or k8s descriptors, ...

Reply:
Now understood, cool